### PR TITLE
Fix: Context state-setting assert fails when `ctx->logits_all` is `false`

### PR DESCRIPTION
**Pull request by ptsochantaris**
If a model context is running with the `logits_all` parameter set to false, then saving and restoring its state via any code path that passes through `llama_set_state_data` will trap at the `GGML_ASSERT` that is tweaked here, because `ctx->logits` will not have been sized to the relevant size at the time (from what I understand, that happens in the lines below).

This PR adds this possibility in the assert logic so that it doesn't trap (however I do hope I have understood what is happening in that code and haven't just "cured a symptom" instead. Happy to close the PR if that's the case :)

### CODE REVIEW

**Comment by cebtenzzre** [comment]
I've been setting logits_all to true when using the new batch API so the state is serialized with a consistent size. That allows me to check if the state is the expected size at load time in order to avoid crashing if the user attempts to load state from a llama.cpp version that predates the batch API.

Once the old API is removed this will be less confusing.

**Comment by ptsochantaris** [comment]
@cebtenzzre You're totally right, in that if we set `logits_all` on the context then there is no crash because, like you say, the capacity of that vector has been set, but the reason I'd prefer not to use that flag in the first place was to avoid generating logits for all tokens while the model runs, which I assume has memory and GPU/CPU overhead? (...like I said though, I may be missing some important info context, in which case apologies if I'm asking possibly naive questions, I'm not an ML expert :))

**Comment by cebtenzzre** [comment]
> I'd prefer not to use that flag in the first place was to avoid generating logits for all tokens while the model runs

logits_all is not honored by llama_decode when you use the new batch API - it respects the value of batch.logits instead. So there's no harm in setting it.

**Comment by ptsochantaris** [comment]
Ah, that's really useful to know, thanks for that! I do indeed use the new batch API 👍

**Comment by ptsochantaris** [comment]
I believe https://github.com/ggerganov/llama.cpp/pull/4820 corrects the issues that this PR would work around, so it's no longer needed. I'll close it 👍
